framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java

```diff
@@ -522,51 +522,80 @@ public JavaExpression visit(NameExpr expr, JavaExpressionContext context) {
                     if (fieldElem != null) {
                         break;
                     }
-                    receiverType = getTypeOfEnclosingClass((DeclaredType) receiverType);
-                    isOriginalReceiver = false;
+                    enclosingTypeOfField =
+                            getTypeOfEnclosingClass((DeclaredType) enclosingTypeOfField);
                 }
             }
-            if (fieldElem != null && fieldElem.getKind() == ElementKind.FIELD) {
-                FieldAccess fieldAccess =
-                        getFieldJavaExpression(fieldElem, receiverExpr, isOriginalReceiver);
-                TypeElement scopeClassElement =
-                        TypesUtils.getTypeElement(fieldAccess.getReceiver().getType());
-                if (!isOriginalReceiver
-                        && !ElementUtils.isStatic(fieldElem)
-                        && ElementUtils.isStatic(scopeClassElement)) {
-                    throw new ParseRuntimeException(
-                            constructJavaExpressionParseError(
-                                    identifier,
-                                    "a non-static field can't be referenced from a static inner class or enum"));
-                }
-                return fieldAccess;
+            if (fieldElem == null) {
+                // field not found.
+                return null;
             }
 
-            return null;
+            // Construct a FieldAccess expression.
+            if (ElementUtils.isStatic(fieldElem)) {
+                Element classElem = fieldElem.getEnclosingElement();
+                JavaExpression staticClassReceiver = new ClassName(ElementUtils.getType(classElem));
+                return new FieldAccess(staticClassReceiver, fieldElem);
+            } else if (receiverExpr instanceof ClassName) {
+                throw new ParseRuntimeException(
+                        constructJavaExpressionParseError(
+                                fieldElem.getSimpleName().toString(),
+                                "a non-static field cannot have a class name as a receiver."));
+            }
+
+            @SuppressWarnings("interning:not.interned") // Checking for exact object
+            boolean fieldDeclaredInReceiverType = enclosingTypeOfField == receiverExpr.getType();
+            // fieldElem is an instance field
+            if (fieldDeclaredInReceiverType) {
+                // It's an instance field declared in the type of receiverExpr.
+                TypeMirror fieldType = ElementUtils.getType(fieldElem);
+                return new FieldAccess(receiverExpr, fieldType, fieldElem);
+            }
+            if (!(receiverExpr instanceof ThisReference)) {
+                throw new ParseRuntimeException(
+                        constructJavaExpressionParseError(
+                                identifier,
+                                "field declared in an outer type cannot be accessed from an inner type"));
```

Comment:
How do we know this is a field declared in an outer type?
In particular, it seems to me that this code could be reached by
```java
class A {
  int f;
}
class B extends A {}

B myB;
... myB.f ...
```

because `f` is not declared in the receiver type and the receiver is not a `ThisReference`.  Am I missing something?

Reply:
In your example, `enclosingTypeOfField == receiverExpr.getType()`, so this code isn't reached.  `resolver.findField(...)` search in supertypes. I added documentation to make this clearer.

Reply:
(Ah, the added documentation is for Resolver.findField.  Yes, that is helpful.)